### PR TITLE
FAPI 2.0 security profile - supporting RFC 9207 OAuth 2.0 Authorization Server Issuer Identification

### LINKED ISSUE #20584: FAPI 2.0 security profile - supporting RFC 9207 OAuth 2.0 Authorization Server Issuer Identification
### Description

[RFC 9207 OAuth 2.0 Authorization Server Issuer Identification](https://www.rfc-editor.org/rfc/rfc9207.html#name-authorization-server-metada) is for preventing [mix-up attack](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19#section-4.4) which may happen if a client treats multiple authorization servers.

### Discussion

_No response_

### Motivation

To strengthen the security of Keycloak, it is valuable for keycloak to support RFC 9207.

Moreover, [FAPI 2.0 security profile](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html), the API security standard specification that OpenID foundation provides [certificate program](https://openid.net/certification/#FAPI2), also requires supporting RFC 9207.

### Details

In authorization code flow, adding "iss" parameter showing an authorization server's identity.

- "iss" parameter value is the same as "iss" field of an access, ID, and refresh token.
- "iss" parameter is always added. According to RFC 6749, a client ignore a parameter that it cannot understand. So, adding "iss" parameter does not affect operations by the current clients using keycloak.

### PR DESCRIPTION
**Pull request by tnorimat**
Closes #20584

### CODE REVIEW

**Comment by pedroigor** [comment]
@tnorimat The mix-up attack being mitigated by this implementation is basically an additional barrier to Keycloak clients, right? 

Most of the time, I see libraries integrating with Keycloak (including our adapters) validating the `iss` claim in both ID and our JWT Access tokens.

It looks more like a OAuth2-specific exploit, right? Is FAPI not forcing clients to use this parameter?

**Comment by mposolda** [comment]
@pedroigor There is some interesting scenario described here https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19#section-4.4.1 .

@pedroigor @tnorimat I think that this is nice addition, but it will be likely still good to mention this in the "Upgrading guide" IMO. The OAuth2 mentions that client should ignore unknown parameters https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2 , however I still have some fear that people will start reporting bugs due their client adapters screaming about "Unknown parameter: iss" ;-) So will be good to have this documented at least. But apart from that, I am not seeing any reason why the parameter should not be included by default and I hope we don't need any additional configuration for that.

@pedroigor Can you think of any issue with including `iss` in the authorization response by default?

**Comment by tnorimat** [comment]
@pedroigor @mposolda FAPI 2.0 requires this feature (#7 of [FAPI 2.0 spec](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html#section-5.3.1.2)). The conformance test of FAPI 2.0 security profile checks this point, so we need this feature to pass the conformance test.

IMO, it is good to add iss parameter as default, and I agree with Marek's proposal for the documentation.
Of course, we add a settings of a client to determine whether iss parameter is added or not, but the number of settings increases.
It is difficult to realize this feature by Client policies because the current client policies cannot add iss parameter to a normal and error authorization response.

**Comment by tnorimat** [comment]
@pedroigor @mposolda Some tests failed because an application using keycloak oidc adapter cannot strip iss parameter when it create redirect_uri parameter of a token request. I stripped iss parameter by [adapter codes of OAuthRequestAuthenticator](https://github.com/keycloak/keycloak/pull/20621/files#diff-f33a1c0a7187111ab3a15d98c6e068d4140579fc50d3cfabb402bda8b24f3eb2) and [ServerRequest](https://github.com/keycloak/keycloak/pull/20621/files#diff-db064681136ab6793c8c65df751b45eb474c43ea0d984b7957ca75e5c7945d3c). However, it seems that they did not work. If possible, could you give me some advices?

**Comment by tnorimat** [comment]
@pedroigor @mposolda I do not know the reason why but erros happend by keycloak-adapter-core were resolved. I will investigate remaining errors in Base IT (5).

**Comment by pedroigor** [comment]
@tnorimat @mposolda I don't see any issue. I was just curious if that affects OpenID Connect clients (not pure OAuth2 clients).

The fact you have the issuer there (and we are using JWT), makes this attack not really possible if your OIDC client (e.g.: quarkus-oidc, Elytron OIDC, or even our legacy adapters) is checking the `iss` from the ID Token.

Adding this change to the release notes makes a lot of sense ...

**Comment by mposolda** [comment]
@pedroigor I think attack may be still possible even for OIDC clients.

If I understand correctly the scenario in https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19#section-4.4.1, the idea is that client receives AuthorizationResponse from "Honest AS", but then it sends the token request to "Attacker AS" (including `code` and `state` etc), which means that attacker will be able to use the `code` to obtain the tokens for "Honest AS". Per my understanding, the `iss` parameter in AuthorizationResponse is a hint for client, which AS to use. So since client see `iss` for "Honest AS", it won't try to send TokenRequest to "attacker AS".

**Comment by pedroigor** [comment]
I think I'm starting to understand how OIDC is impacted after reading https://arxiv.org/abs/1704.08539 and https://arxiv.org/abs/1508.04324v2.

Thanks.

**Comment by jonkoops** [comment]
@pedroigor @mposolda is this something we can land soon, are we blocked on anything specific?

**Review by pedroigor — DISMISSED** [review]
Sorry, for the delay and noise :)

@tnorimat @mposolda LGTM.

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

**Comment by mposolda** [comment]
@tnorimat The PR looks good to me as well, but there are test failures. Maybe rebase on top of latest main can help, but not sure if that's sufficient...

**Review by jonkoops — DISMISSED** [review]
Changes in Keycloak JS look good to me.

**Comment by jonkoops** [comment]
@tnorimat could you give this PR a quick rebase, and possibly fix the tests if they still fail?

**Comment by tnorimat** [comment]
@jonkoops @mposolda 
IMO, it might be better to have add one client configuration for enabling RFC 9207 "iss" parameter because
- A lot of integraiton tests failed because these tests expects that an authorization response does not include "iss" parameter. I am afraid that keycloak users also expect it for their application tests.
- When preparing [FAPI 2.0 security profile as defult](https://github.com/keycloak/keycloak/issues/21181), its profile's executor can enforce the client to RFC 9207 "iss" parameter.

Marek proposed the following options:

1. Add "iss" by some client policy executor. This means that it won't be added by default.
2. Add the "backward compatibility" switch to client similar to "Exclude session_state" parameter. We can just have "Exclude iss" parameter. This means that "iss" parameter would be added by default (similarly like "session_state"), but will be explicitly excluded just if disabled switch on the particular client.

I prefer the option 2, but IMO, it might be good that "iss" parameter would not be added by default because RFC 9207 "iss" parameter is optional. Therefore, it might be good that one client configuration parameter for adding "iss" field is added even I know its drawback that increases the number of client configuration. @mposolda WDYT?

**Comment by mposolda** [comment]
@tnorimat TBH I would prefer to rename the switch to `Exclude Issuer from Authentication Response` . It will be good to be compatible with the switch `Exclude Session State from Authentication Response`, which is kind of similar thing (To me, it is a bit confusing to have inclusion for one of those and have exclusion for the other by default...). If there are any test failures (which AFAIK they were in the previous version of your PR), then maybe you can just disable the `iss` for particular clients in the particular tests, which is also a good prove that this works as expected :-)

Also there may be a need to update admin console as well and add the new switch to backwards compatibility switches.

Also it will be good to add the note to the upgrading guide similar to this https://www.keycloak.org/docs/latest/upgrading/index.html#added-session_state-parameter-to-openid-connect-authentication-response about this switch.

**Comment by tnorimat** [comment]
@mposolda Thank you for the comments and suggestions. I will follow them and review the PR.

**Comment by mposolda** [comment]
@tnorimat Thanks for the updates! Hopefully the failing tests are possible to solve by excluding `iss` or the other way...

For the upgrading guide, hopefully lots of the things can be (almost) copy-pasted from the similar post for "Exclude session_state" parameter from https://www.keycloak.org/docs/latest/upgrading/index.html#added-session_state-parameter-to-openid-connect-authentication-response

**Comment by tnorimat** [comment]
@mposolda Thank you. I will fix failed CI tests and add migration codes (the current client's `exclude iss parameter` setting become ON from default OFF).

**Comment by tnorimat** [comment]
@mposolda It seems that [the PR of `Exclude session_state parameter`](https://github.com/keycloak/keycloak/pull/4928) did not include the migration code (default `OFF` to `ON`). IMO, the PR also need not to do such migration. WDYT?
Also,  I will add documentaiton of migration guide the same as `Exclude session_state parameter`.

**Comment by mposolda** [comment]
@tnorimat My vote is to have `Exclude Issuer from Authentication Response` to be OFF by default. Again - same behaviour like for session_state. Which means that `iss` parameter would be added by default. Administrator can disable `iss` explicitly for particular clients just if it causes the issues for him.

I don't see any issues regarding specifications? As you pointed before - RP should ignore unknown parameters in AuthorizationResponse if it doesn't understand them.

This means that we don't need to do any migration changes (As it is OFF by default for new as well as for existing clients).

**Comment by tnorimat** [comment]
@mposolda Yes, I agree with you.

**Comment by tnorimat** [comment]
@mposolda I revised the PR as you suggested. Could you check it? One CiI action failed (cypress/flake), but it seems that it is not relevant to the PR.

**Review by mposolda — COMMENTED** [review]
@tnorimat Thanks Takashi! I've added 2 minor comments inline.

@andymunro Could you please review the docs to this PR?

@pedroigor Does this OK to you? Is it fine to you to merge this before 22 release or do you prefer to wait after 22 release? For me, it's likely OK to do now (since we have backwards compatibility option), but I prefer to have 2nd opinion too.

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc [hunk 1]
```diff
@@ -357,4 +357,18 @@ Multiple Keycloak CRs may be created in the same namespace and will be managed i
 
 = k8s.keycloak.org/v2alpha1 changes
 
-The condition status field was changed from a boolean to a string for conformance with standard Kubernetes conditions. In the CRD it will temporarily be represented as accepting any content, but it will only ever be a string.  Please make sure any of your usage of this field is updated to expect the values "True", "False", or "Unknown", rather than true or false.
\ No newline at end of file
+The condition status field was changed from a boolean to a string for conformance with standard Kubernetes conditions. In the CRD it will temporarily be represented as accepting any content, but it will only ever be a string.  Please make sure any of your usage of this field is updated to expect the values "True", "False", or "Unknown", rather than true or false.
+
+= Added iss parameter to OAuth 2.0/OpenID Connect Authentication Response
+
+RFC 9207 OAuth 2.0 Authorization Server Issuer Identification specification addes the parameter `iss` in the OAuth 2.0/OpenID Connect Authentication Response for realizing secure authorization responses.
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:364]: Yes, I will fix the typo.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocol.java [hunk 2]
```diff
@@ -278,6 +281,7 @@ public Response authenticated(AuthenticationSessionModel authSession, UserSessio
                 event.error(cpe.getError());
                 new AuthenticationSessionManager(session).removeAuthenticationSession(realm, authSession, true);
                 redirectUri.addParam(OAuth2Constants.ERROR_DESCRIPTION, cpe.getError());
+                redirectUri.addParam(OAuth2Constants.ISSUER, clientSession.getNote(OIDCLoginProtocol.ISSUER));
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocol.java:284]: Yes. I will fix the point.

**Comment by pedroigor** [comment]
I'm not 100% convinced about the new setting but I see it can avoid headaches when upgrading.

**Review by pedroigor — DISMISSED** [review]

**Review by jonkoops — DISMISSED** [review]

**Comment by mposolda** [comment]
@tnorimat The PR looks good to me, Thanks for all the updates!

I think that test failures are not related to this PR, so should be fine.

I have last comment related to docs: Is it possible to move the changes to file changes-23_0_0.adoc instead of changes-22_0_0.adoc? As Keycloak 22 is already released and we unfortunately did not manage to have this in time for 22. Maybe you will need to create the file changes-23_0_0.adoc and add it to changes.adoc.

@andymunro Could you please review the documentation part of this PR? See my previous comment for the migration, so you can ignore this in your review.

**Comment by tnorimat** [comment]
@mposolda I added the file changes-23_0_0.adoc .

**Review by mposolda — DISMISSED** [review]
@tnorimat Thanks! I am approving as this looks good to me.

I think we may have review of the documentation from @andymunro today. Also I hope that failing tests are not a regression of this PR, so I've re-triggered them.

**Review by jonkoops — CHANGES_REQUESTED** [review]
There are some linting issue for the UI code. You probably need to update the dependencies to fix them. Try running `pnpm i && pnpm --filter admin-ui run lint -- --fix` from the `js` directory.

File: docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+RFC 9207 OAuth 2.0 Authorization Server Issuer Identification specification adds the parameter `iss` in the OAuth 2.0/OpenID Connect Authentication Response for realizing secure authorization responses.
+
+In past releases, we did not have this parameter, but now {project_name} adds this parameter by default, as required by the specification.
+
+However, some OpenID Connect / OAuth2 adapters, and especially older {project_name} adapters, may have issues with this new parameter.
+
+For example, the parameter will be always present in the browser URL after successful authentication to the client application.
+In these cases, it may be useful to disable adding the `iss` parameter to the authentication response. This can be done
+for the particular client in the {project_name} admin console, in client details in the section with `OpenID Connect Compatibility Modes`,
```
Comment (andymunro) [inline docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc:11]:
```suggestion
for the particular client in the {project_name} Admin Console, in client details in the section with `OpenID Connect Compatibility Modes`,
```

**Review by andymunro — CHANGES_REQUESTED** [review]
@tnorimat Just one change to request. Thanks.

File: docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc (hunk 3 above)
Comment (tnorimat) [reply docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc:11]: Thank you. I fixed it.

**Comment by mposolda** [comment]
@tnorimat Thanks! We should be good for docs now, so the only missing piece might be to fix the UI test failure. Maybe rebase helps (if you did not already?). Another possibility is to do what @jonkoops suggested in the comment above?

@jonkoops @tnorimat Also I don't know if it is relevant, but when I browse through the files https://github.com/keycloak/keycloak/pull/20621/files, then going bottom to the file OpenIDConnectCompatibilityModes.tsx and I see some message like:
```
 Check failure on line 73 in js/apps/admin-ui/src/clients/advanced/OpenIdConnectCompatibilityModes.tsx
GitHub Actions / Admin UI

Insert `,`
```
Could it be related to UI test failure?

**Comment by tnorimat** [comment]
@mposolda @jonkoops I fixed the tsx file as Marek suggested.

**Review by mposolda — APPROVED** [review]
@tnorimat Seems it helped, UI tests are OK now. Thanks!

@jonkoops Is PR ok for you or is anything still missing from your point of view?

**Review by jonkoops — APPROVED** [review]
@mposolda this all looks good to me 👍

**Comment by mposolda** [comment]
@jonkoops @andymunro @tnorimat Thanks for the reviews and for all the work on this!

**Comment by tnorimat** [comment]
@mposolda Thank you. I will try to work on https://github.com/keycloak/keycloak/issues/21181 . Also, I will re-check whether the current keycloak can pass FAPI 2.0 Impl conformance test.

**Pull request merged**